torch/sparse/_semi_structured_conversions.py

```diff
@@ -1,7 +1,50 @@
 import torch
 
 
-def _sparse_semi_structured_from_dense_cutlass(dense):
+# This is PyTorch implementation of main part of reorder_meta()
+# function, from tools/util/include/cutlass/util/host_reorder.h file
+# of CUTLASS source tree.  Furthermore, CUTLASS template for sparse
+# GEMM decides upon layout of this matrix, and at the moment for the
+# sparse GEMM executed on tensor cores, this is layout described by
+# ColumnMajorInterleaved<2> data structure, in
+# include/cutlass/layout/matrix.h of CUTLASS source tree.  The
+# reordering of meta matrix into meta_reordered matrix calculated
+# according to these segments of CUTLASS code is re-implemented here.
+# Note that this calculation produces offsets for scattering metadata
+# matrix elements into reordered metadata matrix elements (or,
+# equivalently, for gathering reordered metadata matrix element back
+# into metadata matrix elements).
+def _calculate_meta_reordering_scatter_offsets(m, meta_ncols, meta_dtype, device):
```

Comment:
Does the code in this function and the code [here](https://github.com/pytorch/pytorch/pull/116017/files#diff-c7ba2441e5192bccd1becb71844c84d6657506c4abdf2300c3f5b829aad1c552L179-L216) do the same thing - if so is there a reason why we no longer need the "fast" method?

Reply:
Yes, I replaced this complicated code with a direct rewrite (that I actually started from) of CPU code provided in the CUTLASS library doing the conversion.  I found the code related to conversion back to dense tensor time consuming to extend for float32.  The original code was written with the idea that it will be always compiled for Triton, even in eager mode; as we moved away from that approach, my opinion is that it's going to be easier to maintain and eventually extend this simpler code.  Of course, we can always revert; on the other side, I hope CUTLASS guys may provide GPU implementation themselves, in particular as you mentioned earlier that it's going to be there in cuSPARSELt.

Reply:
Question (unrelated): Do we have a set of shapes that would be relevant to run the benchmark script attached above?  The shapes cited in the script are ones that I'm using for while for this purpose, but it would be interesting to benchmark on larger ones.

Reply:
@alexsamardzic That makes sense, I may try to spend a day to write some fast cuSPARSELt torch.compile routines, but long term we're looking to merge in the fast ones @danthe3rd wrote. No need to revert, just wanted to understand the code. 

> Question (unrelated): Do we have a set of shapes that would be relevant to run the benchmark script attached above?

Let me collect this together for all the relevant models (SAM, stable-diffusion, LLaMa) and update the script